cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/KafkaCruiseControl.java

```diff
@@ -923,6 +927,142 @@ public StopProposalResult stopProposalExecution(StopProposalParameters parameter
+   * in a rack-aware, round-robin way.
+   * There are two scenarios that rack awareness property is not guaranteed.
+   * <ul>
+   *   <li> If metadata does not have rack information about brokers, then it is only guaranteed that new replicas are
+   *   added to brokers which currently do not host any replicas of partition.</li>
+   *   <li> If replication factor to set for the topic is larger than number of racks in the cluster and
+   *   skipTopicRackAwarenessCheck is set to true, then rack awareness property is ignored.</li>
+   * </ul>
+   * If partition's current replication factor is larger than target replication factor, remove one or more follower replicas
+   * from the partition.
+   *
+   * @param cluster The metadata of the cluster.
+   * @param replicationFactor The replication factor to set for the topics.
+   * @param topics The topics to apply the change.
+   * @param skipTopicRackAwarenessCheck Whether skip ignore rack awareness property if number of rack in clsuter is less
+   *                                    than target replication factor.
+   * @return Execution proposals to increase replication factor of topics.
+   */
+  private List<ExecutionProposal> maybeUpdateTopicReplicationFactor(Cluster cluster,
+                                                                    int replicationFactor,
+                                                                    Set<String> topics,
+                                                                    boolean skipTopicRackAwarenessCheck) {
+    Map<String, List<Integer>> brokersByRack = new HashMap<>();
+    Map<Integer, String> rackByBroker = new HashMap<>();
+    for (Node node : cluster.nodes()) {
+      // If the rack is not specified, we use the broker id info as rack info.
+      String rack = node.rack() == null || node.rack().isEmpty() ? String.valueOf(node.id()) : node.rack();
+      brokersByRack.putIfAbsent(rack, new ArrayList<>());
+      brokersByRack.get(rack).add(node.id());
+      rackByBroker.put(node.id(), rack);
+    }
+
+    if (replicationFactor > brokersByRack.size()) {
+      if (skipTopicRackAwarenessCheck) {
+        LOG.info(String.format("Target replication factor for topics %s is larger than number of racks in cluster, new replica"
+                               + " maybe added in none rack-aware way.", topics));
+      } else {
+        throw new RuntimeException(String.format("Unable to increase topics %s replica factor to %d since there are only %d " +
+                                                 "racks in the cluster.", topics, replicationFactor, brokersByRack.size()));
+      }
+    }
+
+    List<ExecutionProposal> proposals = new ArrayList<>();
+    for (String topic : topics) {
+      List<String> racks = new ArrayList<>(brokersByRack.keySet());
+      int[] cursors = new int[racks.size()];
+      int rackCursor = 0;
+      for (PartitionInfo partitionInfo : cluster.partitionsForTopic(topic)) {
```

Comment:
Nit: For ease of readability and avoid repetition (in `if` and else `parts`), is it possible to move this logic to separate smaller functions?

Reply:
I feel it is hard to extract the common logic since the only common logic is to check no offline replica and initialize local variable.

Reply:
Is it possible to move the relevant content of the loop to a smaller function -- e.g. something like (potentially with more/less parameters):

```
private Set<ExecutionProposal> updateReplicationFactorFor(PartitionInfo partitionInfo, short replicationFactor) {...}
```
This helps parsing the logic easier and makes it more modular.